frontend/src/lib/components/HgLogView.svelte

```diff
@@ -0,0 +1,134 @@
+                cols[node] = curCol;
+            }
+            entry.col = curCol;
+            const parentCount = parents.length;
+            if (parentCount == 0) {
+                // Nothing to do. Will only happen on the last row.
+            } else if (parentCount == 1) {
+                const parent = parents[0];
+                if (parent in cols) {
+                    // Parent has already been assigned a column, but maybe it can move to the left.
+                    // Therefore, use either its column or the current column, *whichever is lower* (important)
+                    curCol = Math.min(cols[parent], curCol);
+                    cols[parent] = curCol;
+                } else {
+                    cols[parent] = curCol;
+                }
+            } else {
+                let colsAssigned = 0;
+                for (const parent of parents) {
+                    if (parent in cols) {
+                        // Don't touch already-assigned column
+                    } else {
+                        // First new (not seen yet) parent gets current column
+                        // Subsequent ones get one more column to the right
+                        if (colsAssigned == 0) {
+                            cols[parent] = curCol;
+                        } else {
+                            curCol++;
+                            cols[parent] = curCol;
+                        }
+                        colsAssigned++;
+                    }
+                }
+            }
+        }
+    }
+
+    function assignPaths(entries: ExpandedLogEntry[]) {
+        let indices = {} as { [node: string]: number };
+        let paths: [number, number][] = [];
+        for (const entry of entries) {
+            const { node, row } = entry;
+            indices[node] = row;
+        }
+        let curIdx = 0;
+        for (const entry of entries) {
+            const { parents } = entry;
+            for (const parent of parents) {
+                const parentIdx = indices[parent];
+                paths.push([curIdx, parentIdx]);
+            }
+            curIdx++;
+        }
+        return paths;
+    }
+
+    let expandedLog:  ExpandedLogEntry[];
+    $: {
+        expandedLog = json as ExpandedLogEntry[];
+        assignRowsAndColumns(expandedLog);
+        expandedLog = expandedLog.map(e => ({...e, trimmedLog: trimEntry(e.desc)}))
+    };
+
+    function trimEntry(orig: string) {
+        // The [program: version string] part of log entries can get quite long, so let's trim it for the graph
+        return orig.replace(logEntryRe, "");
+    }
+
+    $: circles = expandedLog.map((entry, idx) => ({ row: idx, col: entry.col }));
+    $: paths = assignPaths(expandedLog);
+
+    let heights: number[] = [];
+</script>
+
+<!-- TODO: Create table with log entries, then capture row heights -->
+<div class="horiz">
+<TrainTracks {circles} {paths} rowHeights={heights} />
```

Comment:
if we wrap this in an `#if heights.length == json.length` then it won't show the train tracks until we know the correct height, however that'll then move the table around once it renders the train tracks. Maybe we should do the check inside the TrainTracks component and then we just don't render our svg until we have the heights but we still take up some width.

Reply:
Yeah, inside the TrainTracks component makes sense. I think the check should be `heights.length > 0`, because there might be scenarios where we want to render only a partial set of train tracks, or render a lot of train tracks but only some of the log... and I've verified that Svelte won't do partial updates to the heights array. It'll either be nothing or have the entire set of height values. So `heights.length > 0` will only cause a single re-render.